DatadogSessionReplay/SRExample/SRSnapshotTests/Utils/ImageComparison.swift

```diff
@@ -0,0 +1,249 @@
+        return "Newly-taken snapshot is empty."
+    }
+    guard oldCgImage.width == newCgImage.width, oldCgImage.height == newCgImage.height else {
+        return "Newly-taken snapshot@\(new.size) does not match reference@\(old.size)."
+    }
+    let pixelCount = oldCgImage.width * oldCgImage.height
+    let byteCount = imageContextBytesPerPixel * pixelCount
+    var oldBytes = [UInt8](repeating: 0, count: byteCount)
+    guard let oldData = context(for: oldCgImage, data: &oldBytes)?.data else {
+        return "Reference image's data could not be loaded."
+    }
+    if let newContext = context(for: newCgImage), let newData = newContext.data {
+        if memcmp(oldData, newData, byteCount) == 0 { return nil }
+    }
+    var newerBytes = [UInt8](repeating: 0, count: byteCount)
+    guard
+        let pngData = new.pngData(),
+        let newerCgImage = UIImage(data: pngData)?.cgImage,
+        let newerContext = context(for: newerCgImage, data: &newerBytes),
+        let newerData = newerContext.data
+    else {
+        return "Newly-taken snapshot's data could not be loaded."
+    }
+    if memcmp(oldData, newerData, byteCount) == 0 { return nil }
+    if precision >= 1, perceptualPrecision >= 1 {
+        return "Newly-taken snapshot does not match reference."
+    }
+    if perceptualPrecision < 1 {
+        return perceptuallyCompare(
+            CIImage(cgImage: oldCgImage),
+            CIImage(cgImage: newCgImage),
+            pixelPrecision: precision,
+            perceptualPrecision: perceptualPrecision
+        )
+    } else {
+        let byteCountThreshold = Int((1 - precision) * Float(byteCount))
+        var differentByteCount = 0
+        for offset in 0..<byteCount {
+            if oldBytes[offset] != newerBytes[offset] {
+                differentByteCount += 1
+            }
+        }
+        if differentByteCount > byteCountThreshold {
+            let actualPrecision = 1 - Float(differentByteCount) / Float(byteCount)
+            return "Actual image precision \(actualPrecision) is less than required \(precision)"
+        }
+    }
+    return nil
+}
+
+private func context(for cgImage: CGImage, data: UnsafeMutableRawPointer? = nil) -> CGContext? {
+    let bytesPerRow = cgImage.width * imageContextBytesPerPixel
+    guard
+        let colorSpace = imageContextColorSpace,
+        let context = CGContext(
+            data: data,
+            width: cgImage.width,
+            height: cgImage.height,
+            bitsPerComponent: imageContextBitsPerComponent,
+            bytesPerRow: bytesPerRow,
+            space: colorSpace,
+            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
+        )
+    else { return nil }
+
+    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: cgImage.width, height: cgImage.height))
+    return context
+}
+
+func perceptuallyCompare(_ old: CIImage, _ new: CIImage, pixelPrecision: Float, perceptualPrecision: Float) -> String? {
```

Comment:
Same here: `DDAssertPerceptualInageComparison`, or something like that?

Reply:
For that one, I didn't wrap it into separate `DDAssert*`. We don't use this method independently - it's part of `DDAssertSnapshotEquals(...)` logic 👍.